diff --git a/gemrb/core/Scriptable/Actor.cpp b/gemrb/core/Scriptable/Actor.cpp
index bd364d8..5f87dac 100644
--- a/gemrb/core/Scriptable/Actor.cpp
+++ b/gemrb/core/Scriptable/Actor.cpp
@@ -4187,6 +4187,7 @@ void Actor::CheckCleave()
 static void ChunkActor(Actor* actor)
 {
 	const Map* map = actor->GetCurrentArea();
+	if (!map) return;
 	if (!map->IsVisible(actor->Pos)) return; // protect against ctrl-shift-y
 
 	// TODO: play chunky animation / particles #128
diff --git a/gemrb/plugins/GUIScript/GUIScript.cpp b/gemrb/plugins/GUIScript/GUIScript.cpp
index 9c67d79..8e6a6c8 100644
--- a/gemrb/plugins/GUIScript/GUIScript.cpp
+++ b/gemrb/plugins/GUIScript/GUIScript.cpp
@@ -3461,7 +3461,9 @@ static PyObject* GemRB_Button_SetPLT(PyObject* self, PyObject* args)
 	ieDword col[8] { ieDword(-1), ieDword(-1), ieDword(-1), ieDword(-1), ieDword(-1), ieDword(-1), ieDword(-1), 0 };
 	Actor::stat_t keys[8] { IE_METAL_COLOR, IE_MINOR_COLOR, IE_MAJOR_COLOR, IE_SKIN_COLOR, IE_LEATHER_COLOR, IE_ARMOR_COLOR, IE_HAIR_COLOR };
 	for (int i = 0; i < 8; ++i) {
-		PyObject* obj = PyDict_GetItem(colors, PyLong_FromLong(keys[i]));
+		PyObject* key = PyLong_FromLong(keys[i]);
+		PyObject* obj = PyDict_GetItem(colors, key);
+		Py_DecRef(key);
 		if (obj) {
 			col[i] = ieDword(PyLong_AsLong(obj));
 		}
