diff --git a/gemrb/GUIScripts/GUIRECCommon.py b/gemrb/GUIScripts/GUIRECCommon.py
index e77b86f..dc45989 100644
--- a/gemrb/GUIScripts/GUIRECCommon.py
+++ b/gemrb/GUIScripts/GUIRECCommon.py
@@ -20,7 +20,6 @@ import collections
 import GemRB
 import GameCheck
 import GUICommon
-import PaperDoll
 import Portrait
 from GUIDefines import *
 from ie_stats import *
@@ -88,6 +87,7 @@ def OpenCustomizeWindow ():
 		SoundButton.SetState (IE_GUI_BUTTON_DISABLED)
 
 	if not GameCheck.IsIWD2():
+		import PaperDoll
 		ColorButton = CustomizeWindow.GetControl (2)
 		ColorButton.SetText (10646)
 		stats = PaperDoll.ColorStatsFromPC (pc)
